### PR TITLE
POC for dynamic routing

### LINKED ISSUE #1101: Update backend to handle dynamic frontend routes
Currently the backend doesn't handle dynamic frontend routes. A POC draft pr has most of the work implemented to support them (#1033). It would be great to get this across the finished line so routing works as expected when using the frontend.

### PR DESCRIPTION
**Pull request by TheAndrewJackson**
# Purpose
This PR implements a routing behavior that makes it possible for the webserver to serve dynamic urls for the frontend. It works by scanning the build directory and building out all of the routes dynamically based on all of the `html` files and directories  within it. This works by taking advantage of the fact that the frontend routes have a 1:1 match with the files and directories in the build folder.

# Changes
- Update webserver to build out dynamic routes.

# Checklist
- [x] Update [`CHANGELOG.md`](https://github.com/ethyca/fidesops/blob/main/CHANGELOG.md) file
  - [x] Merge in main so the most recent `CHANGELOG.md` file is being appended to
  - [x] Add description within the `Unreleased` section in an appropriate category. Add a new category from the list at the top of the file if the needed one isn't already there.
  - [x] Add a link to this PR at the end of the description with the PR number as the text. example: [#1](https://github.com/ethyca/fidesops/pull/1)
- [ ] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md).
- If docs updated (select one):
  - [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  - [ ] documentation issue created (tag docs-team to complete issue separately)
- [ ] Good unit test/integration test coverage
- [ ] This PR contains a DB migration. If checked, the reviewer should confirm with the author that the [down_revision correctly references the previous migration](https://ethyca.github.io/fidesops/development/contributing_details/#alembic-migrations) before merging
- [ ] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket

Fixes #1101

### CODE REVIEW

**Review by allisonking — COMMENTED** [review]
nice solution!! I left some comments, but overall wasn't sure how thorough you wanted me to be on this review or if this is still an early stage thing. seems promising, though! 

out of curiosity, did you ever try the solutions suggested in https://github.com/ethyca/fides/issues/942? I haven't tried them, but it'd be cool if there's a solution on the nextjs side of things (I'm of the opinion that this is a nextjs problem, since regular ol' servers can't handle this either)

File: src/fidesops/main.py [hunk 1]
```diff
@@ -130,6 +131,17 @@ def prepare_and_log_request(
 WEBAPP_INDEX = WEBAPP_DIRECTORY / "index.html"
 
 if config.admin_ui.enabled:
+    route_file_map = {}
+
+    def generate_route_file_map() -> None:
```
[the remark this answers is not in the document]
Comment (TheAndrewJackson) [reply src/fidesops/main.py:143]: we can. docstring added :)

File: src/fidesops/main.py [hunk 2]
```diff
@@ -130,6 +131,17 @@ def prepare_and_log_request(
...
+        pattern = r"\[.+]"
+        for filepath in WEBAPP_DIRECTORY.glob("**/*.html"):
+            if "index" not in str(filepath):
```
[the remark this answers is not in the document]
Comment (TheAndrewJackson) [reply src/fidesops/main.py]: It is. I didn't realize you could compare pathlib objects. I updated it to use `WEBAPP_INDEX`. Good find!

File: src/fidesops/main.py [hunk 3]
```diff
@@ -130,6 +131,17 @@ def prepare_and_log_request(
...
+                path = re.sub(
+                    pattern, ".+", str(filepath.relative_to(WEBAPP_DIRECTORY))[:-5]
+                )
+                rule = re.compile(r"^" + path + "/?$")
+                route_file_map[rule] = FileResponse(str(filepath.relative_to(".")))
```
[the remark this answers is not in the document]
Comment (TheAndrewJackson) [reply src/fidesops/main.py:171]: It did sort of handle nested paths. It was looking for anything that was in between 2 brackets `[]`. I updated the routing regex to be more picky and it has code to specifically handle normal and nested routes now. It should handle things more explicitly. thanks for pointing that out.

**Comment by TheAndrewJackson** [comment]
@allisonking Thanks for the feedback! I took a look at `exportPathMap` and it seems like it won't work. It looks like it's used to override the automatic routes that are made based on the filesystem. 

The documentation says
> `exportPathMap` allows you to specify a mapping of request paths to page destinations, to be used during export.

and

> For more advanced scenarios, you can define a parameter called [exportPathMap](https://nextjs.org/docs/api-reference/next.config.js/exportPathMap) in your [next.config.js](https://nextjs.org/docs/api-reference/next.config.js/introduction) file to configure exactly which pages will be generated

From my understanding it seems like we would end up with the same static build but the routes and files that are generated would be based on the `exportPathMap` config instead of the file system.

**Review by chriscalhoun1974 — APPROVED** [review]

**Pull request merged**